README.md
# Character-Creator
Character Creator

| https://www.notkal.com/Character-Creator | [Issues](https://github.com/not-the/Character-Creator/issues) | [Projects](https://github.com/users/not-the/projects/3) |
|---|---|---|


changelog.txt
v11   Nov 21, 2023
- Fixed extra padding below footer depending on what tab you had open
- Updated nav bar

v10   Nov 18, 2022
- Fixed the randomize option not updating which parts appear to be selected
- Added favicon, fixed footer padding
- Fixed Bill preset having the wrong hair
- Fixed Charles' eyes not recoloring the correct pixels
- New presets: Ugly Sweater Bill, Business Bill, Baker Bill, Grandma Belle, Safety Greg
- New parts:
    Background: Cloudless
    Hair: Belle (White)
    Body: Ugly Sweater, Blue Suit, High Vis, Chef, Sweater
    Hat: Chef hat
    Accessory: Glasses (alt)

v9   Nov 18, 2022
- Fixed Greg preset not having his midground
- Added new presets: Fancy Bill, Biker Bill, Dollar Bill
- New parts:
    Body: Leather Jacket

v8   Nov 10, 2022
- Reduced the odds of getting a foreground when clicking the randomize button
- Gave the tabs a button hover effect

v7   Nov 10, 2022
- Removed redundant CSS
- Hats is no longer plural

v6   Oct 11, 2022
- Arrow keys will now reposition the last element you edited
- Randomize button
- Preset dropdown
- Added Jared parts
- Added JJCVIP to the credits
- Fixed hat/hair offset not being calculated properly
- New parts:
    Body: Suit, Jokester, Black shirt
    Hats: Charles XL, Top Hat
    Accessories: Bow tie

v5   Oct 11, 2022
- Hats and hair are now separate
- Added a View on Github button
- Fixed rapid taps on position buttons zooming the page on mobile
- Reset buttons on color pickers now works
- Resetting positioning is now an option
- New parts:
    6 new backgrounds: Night alt, Space, Brick wall, Brick wall blue, Mushroom land, Classic (snow)
    1 new miground: Snow
    7 new hair types: Bill, Bill messy, Belle, Greg, Charles, Carl, Hair loss
    2 new noses: Nose 8, Nose 9
    1 new hat: Hat stack
    2 new accessories: Glasses, Dollar necklace

v4   Oct 8, 2022
- Added positioning options for eyes, nose, mouth, and accessories
- Fixed the width of the parts panel changing sometimes
- No more floating heads, bodies of different heights will now move other elements to match
- Added 4 new mouths

v3   Oct 8, 2022
- Complete CSS overhaul, now matches rest of the website

v2   Oct 7, 2022
- Added a "none" option to the following part types: midground, eyes, mouth, nose, accesory
- New "Foreground" part type
- New parts:
    Carrot Pile background, midground, and foregrounds
    Themery background2123
    Classic background
    Custom Sky background
    Solid Color Background
    Textured Color background
    Bill Buddy foreground
- Gave more parts color customization
    Bill headwear
    Sunglasses accessory
- "Before" custom color mode
- Added a save image button
- "Carrot Clicker" in the title is now a link
- Parts with color options will now have an icon indicating that
- Fixed the selected part not getting highlighted when clicked sometimes
- Fixed the selected part not appearing as such on page load
- Fixed parts not wrapping when larger than the container

v1   Oct 7, 2022
- Fixed an issue where the canvas would flicker when a part is equipped
- Added a bunch of parts and seperated eyes, nose, mouth, and hat from the heads
- Number keys let you change tabs
- Misc CSS fixes

index.js
// Canvas
var c = dom('canvas');
var ctx;
var src_size = 32;
var size = c.width;
var times = size / src_size; // Canvas is 16x larger than assets

let images_loaded = false;

// Data
// In order, back to front
const parts = {
    background: [
        {
            id: 'sky',
            color_default: '#000000',
            color_pixels: [],
        },
        {
            id: 'cloudy',
            color_default: '#000000',
            color_pixels: [],
        },
        {
            id: 'dusk',
            color_default: '#000000',
            color_pixels: [],
        },
        {
            id: 'library',
            color_default: '#000000',
            color_pixels: [],
        },
        // {
        //     id: 'solid color',
        //     color_default: '#888888',
        //     color_pixels: [
        //         [0, 0, 32, 32]
        //     ],
        // },
    ],
    midground: [
        'blacksmith',
        'shop',
    ],
    body: [
        {
            id: 'bill',
            color_default: '#3851bf',
            color_pixels: [
                [9, 22, 1, 10],
                [22, 22, 1, 10],
            ],
        },
        {
            id: 'belle',
            color_default: '#3851bf',
            color_pixels: [
                [9, 22, 1, 10],
                [22, 22, 1, 10],
            ],
        },
        {
            id: 'greg',
            color_default: '#3851bf',
            color_pixels: [
                [10, 22, 1, 10],
                [20, 22, 1, 10],
            ],
        },
        {
            id: 'charles',
            color_default: '#3851bf',
            color_pixels: [
                [14, 22, 1, 7],
                [15, 22, 1, 8],
                [15, 24, 1, 2, '#000000', 0.12],
                [15, 27, 1, 1, '#000000', 0.12],
            ],
        },
        {
            id: 'carl',
            color_default: '#000000',
            color_pixels: [],
        },
    ],
    head: [
        'bill',
        'belle',
        'greg',
        'charles',
        'carl',
    ],
    eyes: [
        {
            id: 'bill',
            color_default: '#0db8ff',
            color_pixels: [
                [11, 10],
                [20, 10],
            ],
        },
        {
            id: 'belle',
            color_default: '#88e543',
            color_pixels: [
                [11, 10],
                [20, 10],
            ],
        },
        {
            id: 'greg',
            color_default: '#ab5309',
            color_pixels: [
                [11, 10],
                [21, 10],
            ],
        },
        {
            id: 'charles',
            color_default: '#e7a5fb',
            color_pixels: [
                [12, 12],
                [18, 12],
            ],
        },
    ],
    mouth: [
        'bill',
        'belle',
        'greg',
        'charles',
        'carl',
    ],
    nose: [
        'nose 1', // bill belle greg
        'nose 2',
        'nose 3',
        'nose 4',
        'nose 5',
        'nose 6',
        'nose 7',
        'carl',
    ],
    headwear: [
        'none',
        'bill',
        'belle',
        'greg',
    ],
    accessory: [ // #44474c
        {
            id: 'sunglasses',
            color_default: '#3851bf',
            color_pixels: [
                [14, 22, 1, 7],
                [15, 22, 1, 8],
                [15, 24, 1, 2, '#000000', 0.12],
                [15, 27, 1, 1, '#000000', 0.12],
            ],
        },
        {
            id: 'scarf',
            color_default: '#3851bf',
            color_pixels: [],
        },
        {
            id: 'monocle',
            color_default: '#3851bf',
            color_pixels: [],
        },
    ],
}
const part_data = {
    background: {
        // color: true,
    },
    midground: {},

    body: {
        color: true,
    },
    head: {},
    eyes: {
        color: true,
        color_default: '#0db8ff',
    },
    mouth: {},
    nose: {},
    headwear: {},
    accessory: {},
}
var user = {
    part: {
        background: 0,
        midground:  undefined,

        body:       0,
        head:       0,
        eyes:       0,
        mouth:      0,
        nose:       0,
        headwear:   1,
        accessory:  undefined,
    },
    color: {}
}
let keys = Object.keys(parts);

/** Get image SRC */
function getSRC(type, name) { return `./assets/${type}/${name}.png`}
/** Get item filename */
function itemID(item) { return typeof item == 'object' ? item.id : item; }

/** Populate parts list(s) */
function populate(type) {
    console.log(`Populating: ${type}`);
    let html = '<div class="flex">';
    let list = parts[type];
    for(i = 0; i < list.length; i++) {
        let item = list[i];
        let name = itemID(item);
        html +=
        `
        <figure class="part part_${type}" id="part_${type}_${name}" onclick="equip('${type}', ${i})">
            <img src="./assets/${type}/${name}.png" alt="${name}" class="part_img">
            <figcaption>${capitalizeFL(name)}</figcaption>
        </figure>
        `;
    }

    let container = dom(`parts_${type}`);
    if(container == null) return console.warn('null element');
    html += `</div>`;
    container.innerHTML = html;

    // Color picker
    if(part_data[type].color == true) {
        container.innerHTML += `
        <div class="customize_item flex">
            <h4>${capitalizeFL(type)} Color:</h4>
            <input type="color" name="${type}_color" id="${type}_color">
        </div>`;
        dom(`${type}_color`).addEventListener('input', event => {
            user.color[type] = event.srcElement.value;
            fullDraw();
        });
    }
}

/** Equip part */
function equip(type, num) {
    console.log(type, num);
    if(user.part[type] == num) return;
    user.part[type] = num;
    document.querySelectorAll(`.part_${type}`).forEach(element => element.classList.remove('active'));
    dom(`part_${type}_${parts[type][num]}`)?.classList.add('active');
    fullDraw();
}

/** Draw all selected */
function fullDraw() {
    for(i = 0; i < keys.length; i++) {
        let type = keys[i];
        let num = user.part[type];
        let item = parts[type][num];
        let name = itemID(item); // Get ID
        draw(type, name, num);
        if(part_data[type].color) {
            drawColor(type, user.part[type]);
        }
    }
    images_loaded = true;
}

/** Draw */
function draw(type, name, num, offset) {
    // console.log(`Drawing ${type}: ${name}`);
    if(type == undefined || name == undefined) return console.warn(`Invalid part type or name. Type: ${type} Name: ${name}`);

    let [x, y] = offset || [0, 0];
    x*=times;
    y*=times;

    // Get image
    var img = new Image();
    img.src = getSRC(type, name);

    // Draw
    if(images_loaded) d();
    else img.onload = d;

    if(part_data[type].color == true) drawColor(type, num);

    function d() { ctx.drawImage(img, x, y, size, size); }
}
/** Draw rectangle */
function drawColor(type, num) {
    let pixels = parts[type][num].color_pixels;
    let color = user['color'][type];
    if(color == undefined) return;

    for(let i = 0; i < pixels.length; i++) {
        let pix = pixels[i];
        let [x, y, w, h, color_override, alpha] = pix;
        console.log(pix);
        ctx.fillStyle = color_override || color || 'gray';
        ctx.globalAlpha = alpha || 1;
        ctx.fillRect(x*times, y*times, (w||1)*times, (h||1)*times);
        ctx.globalAlpha = 1;
    }
}

// UI
var current_tab = 'background';
/** Tab */
function tab(choice, state=true) {
    if(choice == current_tab && state) return;

    store('character_creator_tab', choice);

    let button = dom(`tab_${choice}`);
    let content = dom(`parts_${choice}`);
    style(button, 'active_tab', state);
    style(content, 'active_tab', state);

    if(state) tab(current_tab, false);
    current_tab = choice;
}


// Keyboard shortcuts
document.addEventListener('keydown', event => {
    let key = Number(event.key);
    if(key == 0) key = 10;
    if(key != NaN) tab(keys[key - 1]);
})


// On page load
window.onload = () => {
    for(let i = 0; i < keys.length; i++) {
        try { populate(keys[i]); }
        catch (error) { console.error(error); }
        if(i > 32) return console.error("Killing for loop: exceeded 32");
    }

    // Setup
    ctx = c.getContext("2d");
    ctx.imageSmoothingEnabled = false;
    fullDraw();

    // localStorage
    // let stored_tab = store('character_creator_tab');
    // if(stored_tab != null) tab(stored_tab, false);
}

data.js
/** Parts data. In order from back to front */
const parts = {
    background: [
        'sky',
        'cloudy',
        'dusk',
        'library',
        'themery',
        'carrot pile',
        'classic',
        'night alt',
        'space',
        'brick wall',
        'brick wall blue',
        'mushroom land',
        'classic (snow)',
        'cloudless',
        {
            id: 'custom sky',
            color_default: '#888888',
            color_pixels: [
                [0, 0, 32, 32]
            ],
            color_mode: 'before',
        },
        {
            id: 'solid color',
            color_default: '#888888',
            color_pixels: [
                [0, 0, 32, 32]
            ],
        },
        {
            id: 'textured color',
            color_default: '#888888',
            color_pixels: [
                [0, 0, 32, 32]
            ],
            color_mode: 'before',
        },
    ],
    midground: [
        'none',
        'blacksmith',
        'shop',
        'carrot pile',
        'snow',
    ],
    body: [
        {
            id: 'bill',
            color_default: '#3851bf',
            color_pixels: [
                [9, 22, 1, 10],
                [22, 22, 1, 10],
            ],
        },
        {
            id: 'belle',
            color_default: '#3851bf',
            color_pixels: [
                [9, 22, 1, 10],
                [22, 22, 1, 10],
            ],
        },
        {
            id: 'greg',
            color_default: '#3851bf',
            color_pixels: [
                [10, 22, 1, 10],
                [20, 22, 1, 10],
            ],
        },
        {
            id: 'charles',
            color_default: '#3851bf',
            color_pixels: [
                [14, 20, 1,  9],
                [15, 20, 1, 10],
                [14, 20, 1,  1, '#000000', 0.12],
                [15, 20, 1,  2, '#000000', 0.12],
                [15, 24, 1,  2, '#000000', 0.12],
                [15, 27, 1,  1, '#000000', 0.12],
            ],
            head_offset: {x:0, y:-2},
        },
        'carl',
        'jared',
        {
            id: 'suit',
            color_default: '#202224',
            color_pixels: [
                [13, 22, 2, 3],
                [15, 23, 1, 1],
                [16, 22, 2, 3],
            ],
        },
        'jokester',
        'black shirt',
        'leather jacket',
        'ugly sweater',
        'blue suit',
        'high vis',
        'chef',
        'sweater',
    ],
    head: [
        'bill',
        'belle',
        'greg',
        {
            id: 'charles',
            headwear_offset: {x:0, y:3},
            // filter: 'brightness(50%)',
        },
        {
            id: 'carl',
            headwear_offset: {x:0, y:2}
        },
        {
            id: 'jared',
            headwear_offset: {x:0, y:2}
        },
    ],
    hair: [
        'none',
        'bill',
        'bill messy',
        'belle',
        'greg',
        'charles',
        'carl',
        'jared',
        {
            id: 'hair loss',
            color_default: '#594326',
            color_pixels: [
                [7, 8, 2,  1],
                [6, 9, 2,  1],

                [11, 8, 1,  1],
                [20, 8, 2,  1],

                [23, 8, 2,  1],
                [24, 9, 2,  1],
            ],
        },
        'belle (white)',
    ],
    mouth: [
        'none',
        'bill',
        'belle',
        'greg',
        'charles',
        'carl',
        'jared',
        'mouth 1',
        'mouth 2',
        'mouth 3',
        'mouth 4',
    ],
    nose: [
        'none',
        'nose 1', // bill belle greg
        'nose 2',
        'nose 3',
        'nose 4',
        'nose 5',
        'nose 6',
        'nose 7',
        'nose 8',
        'nose 9',
        'carl',   // carl
        'jared',  // jared
    ],
    eyes: [
        'none',
        {
            id: 'bill',
            color_default: '#0db8ff',
            color_pixels: [
                [11, 10],
                [20, 10],
            ],
        },
        {
            id: 'belle',
            color_default: '#88e543',
            color_pixels: [
                [11, 10],
                [20, 10],
            ],
        },
        {
            id: 'greg',
            color_default: '#ab5309',
            color_pixels: [
                [11, 10],
                [21, 10],
            ],
        },
        {
            id: 'charles',
            color_default: '#e7a5fb',
            color_pixels: [
                [12, 14],
                [18, 14],
            ],
        },
        {
            id: 'carl',
            color_default: '#e7a5fb',
            color_pixels: [
                [10, 13],
                [19, 13],
            ],
        },
        {
            id: 'jared',
            color_default: '#b441b1',
            color_pixels: [
                [12, 12],
                [17, 12],
            ],
        },
    ],
    hat: [
        'none',
        {
            id: 'bill',
            color_default: '#ab3a47',
            color_pixels: [
                [ 8,  6, 16, 1],
                [10,  6, 2, 1, '#ffffff', 0.25],
                [14, 6, 1, 1, '#ffffff', 0.25],
            ],
        },
        'belle',
        'greg',
        'charles',
        'charles xl',
        'carl',
        'jared',
        {
            id: 'top hat',
            color_default: '#292c30',
            color_pixels: [
                [ 9,  5, 14, 1],
                [10,  5,  5, 1, '#ffffff', 0.15],
                [16,  5,  2, 1, '#ffffff', 0.15],
                [21,  5,  1, 1, '#ffffff', 0.15],
            ],
        },
        'biker helmet',
        'hat stack',
        'chef hat',
    ],
    accessory: [ // #44474c
        'none',
        'jared beard',
        {
            id: 'glasses',
            color_default: '#3851bf',
            color_pixels: [
                [ 9, 10, 1, 2],
                [10, 12, 3, 1],
                [13, 10, 1, 2],

                [ 18, 10, 1, 2],
                [19, 12, 3, 1],
                [22, 10, 1, 2],
            ],
        },
        'glasses (alt)',
        {
            id: 'sunglasses',
            color_default: '#3851bf',
            color_pixels: [
                [ 9, 10, 5, 2],
                [10, 12, 3, 1],
                [11, 10, 2, 1, '#ffffff', 0.5],
                [10, 11, 2, 1, '#ffffff', 0.5],
                [10, 12, 1, 1, '#ffffff', 0.5],

                [18, 10, 5, 2],
                [19, 12, 3, 1],
                [20, 10, 2, 1, '#ffffff', 0.5],
                [19, 11, 2, 1, '#ffffff', 0.5],
                [19, 12, 1, 1, '#ffffff', 0.5],
            ],
        },
        {
            id: 'scarf',
            // color_default: '#464646',
            // color_pixels: [
            //     [0, 2, 2, 2],
            //     [ 8, 20, 4, 3],
            //     [9, 21, 14, 3],
            //     [ 20, 20, 4, 3],
            // ],
        },
        {
            id: 'monocle',
            // color_default: '#3851bf',
            // color_pixels: [],
        },
        'dollar necklace',
        {
            id: 'bow tie',
            color_default: '#202224',
            color_pixels: [
                [13, 22, 2, 3],
                [15, 23, 1, 1],
                [16, 22, 2, 3],
            ],
        },
    ],
    foreground: [
        'none',
        'carrot pile',
        'bill buddy',
    ],
}

/** What configuration options are available and what they default to */
const part_data = {
    background: {
        color: true,
        color_default: '#888888',
    },
    midground: {},

    body: {
        color: true,
    },
    head: {},
    hair: {
        color: true,
        position: true,
    },
    eyes: {
        color: true,
        color_default: '#0db8ff',
        position: true,
    },
    mouth: {
        position: true,
    },
    nose: {
        position: true,
    },
    hat: {
        color: true,
        position: true,
    },
    accessory: {
        color: true,
        position: true,
        allow_multiple: true,
    },
    foreground: {},
}




/** Presets */
const presets = {
    'Bill': {
        background: 0,
        midground:  0,

        body:       0,
        head:       0,
        hair:       8,
        eyes:       1,
        mouth:      1,
        nose:       1,
        hat:        1,
        accessory:  0,
        foreground: 0,
    },
    'Belle': {
        background: 0,
        midground:  0,

        body:       1,
        head:       1,
        hair:       3,
        eyes:       2,
        mouth:      2,
        nose:       1,
        hat:        2,
        accessory:  0,
        foreground: 0,
    },
    'Greg': {
        background: 0,
        midground:  1,

        body:       2,
        head:       2,
        hair:       4,
        eyes:       3,
        mouth:      3,
        nose:       1,
        hat:        3,
        accessory:  0,
        foreground: 0,
    },
    'Charles': {
        background: 3,
        midground:  0,

        body:       3,
        head:       3,
        hair:       5,
        eyes:       4,
        mouth:      4,
        nose:       2,
        hat:        4,
        accessory:  0,
        foreground: 0,
    },
    'Carl': {
        background: 0,
        midground:  0,

        body:       4,
        head:       4,
        hair:       6,
        eyes:       5,
        mouth:      5,
        nose:      10,
        hat:        6,
        accessory:  0,
        foreground: 0,
    },
    'Jared': {
        background: 0,
        midground:  2,

        body:       5,
        head:       5,
        hair:       7,
        eyes:       6,
        mouth:      6,
        nose:      11,
        hat:        7,
        accessory:  1,
        foreground: 0,
    },
    'Fancy Bill': {
        "background": 0,
        "midground": 0,
        "body": 6,
        "head": 0,
        "hair": 1,
        "eyes": 1,
        "mouth": 1,
        "nose": 1,
        "hat": 8,
        "accessory": 0,
        "foreground": 0
    },
    'Biker Bill': {
        "background": 1,
        "midground": 0,
        "body": 9,
        "head": 0,
        "hair": 1,
        "eyes": 1,
        "mouth": 1,
        "nose": 1,
        "hat": 9,
        "accessory": 0,
        "foreground": 0
    },
    'Dollar Bill': {
        "background": 9,
        "midground": 0,
        "body": 9,
        "head": 0,
        "hair": 2,
        "eyes": 1,
        "mouth": 1,
        "nose": 1,
        "hat": 0,
        "accessory": 7,
        "foreground": 0
    },
    'Ugly Sweater Bill': {
        "background": 13,
        "midground": 4,
        "body": 10,
        "head": 0,
        "hair": 1,
        "eyes": 1,
        "mouth": 1,
        "nose": 1,
        "hat": 0,
        "accessory": 0,
        "foreground": 0
    },
    'Business Bill': {
        "background": 0,
        "midground": 0,
        "body": 11,
        "head": 0,
        "hair": 1,
        "eyes": 1,
        "mouth": 1,
        "nose": 1,
        "hat": 0,
        "accessory": 0,
        "foreground": 0
    },
    'Baker Bill': {
        "background": 0,
        "midground": 0,
        "body": 13,
        "head": 0,
        "hair": 8,
        "eyes": 1,
        "mouth": 1,
        "nose": 1,
        "hat": 11,
        "accessory": 0,
        "foreground": 0
    },
    'Grandma Belle': {
        "background": 0,
        "midground": 0,
        "body": 14,
        "head": 1,
        "hair": 9,
        "eyes": 2,
        "mouth": 2,
        "nose": 1,
        "hat": 0,
        "accessory": 3,
        "foreground": 0
    },
    'Safety Greg': {
        "background": 0,
        "midground": 0,
        "body": 12,
        "head": 2,
        "hair": 4,
        "eyes": 3,
        "mouth": 3,
        "nose": 1,
        "hat": 3,
        "accessory": 0,
        "foreground": 0
    },
}